### PR TITLE
general code cleanup to allow faster broadcasting implementations

### PR DESCRIPTION
**Pull request by avik-pal**
## Main Description

This doesn't really attempt to make things faster. We do see speed improvements from:

- `broadcast`-ing rewritten as loops for CPUs for uniform broadcasting
- more rrules for common operations
- general type-stability improvements

~Mostly to check why https://github.com/LuxDL/LuxLib.jl/pull/85/ is stalling CUDA~ -- fixed now.

We get rid of FastBroadcast.jl (for now) and replace it with loops for the CPU case. The idea here is to bring it back in a later PR, which solely focuses on performance in cases where we cannot use LoopVectorization.

My intuition was Enzyme would be great for loops but current Zygote with rrules absolutely destroys it for some of the operations -- at least `dropout` I checked. We need to add rules to help out enzyme here (especially now that it is the recommended autodiff backend).

## TODOs

- [x] dropout
  - [x] impl
  - [x] test type stability
  - [x] GPU tests fail due to evals for type stability
- [x] downstream testing
  - [x] type stability regression
  - [x] enzyme support regression
- [x] normalization impl update
- [x] affine normalize fuse operations -- layernorm has a very general structure making it very hard to write a loop for it.
  - [x] groupnorm
  - [x] ~batchnorm~ -- deferred to later
  - [x] ~instancenorm~ -- deferred to latersome refactor
- [x] rework the fused impls using `get_device_type` to minimize code in extensions.
  - [x] fused_conv
  - [x] fused_dense
- [x] testing type stability
  - [x] BN
  - [x] GN
  - [x] IN
  - [x] LN
- [x] Add Enzyme Tests
  - [x] dropout
  - [x] alpha_dropout
  - [x] conv
  - [x] dense
  - [x] batch_norm
  - [x] group_norm
  - [x] layer_norm
  - [x] instance_norm
- [x] rework broadcasting function and replace more smaller components directly
- [x] doc strings for bias_activation

## Before Merging

- [x] remove installing master for packages from runtests
- [x] remove parallel testing on gpus
- [x] skip enzyme float16 on windows
- [x] remove `-g2` from downstream testing

### CODE REVIEW

**Comment by avik-pal** [comment]
@wsmoses I was adding more coverage for enzyme in the backend. https://github.com/LuxDL/LuxLib.jl/actions/runs/9986274451/job/27598503161?pr=86#step:6:400 suddenly showed up. Do you know what this means?

**Comment by avik-pal** [comment]
Seems like a uniquely windows + Float16 issue that I don't care too much about (and lack the ability to reproduce locally)
